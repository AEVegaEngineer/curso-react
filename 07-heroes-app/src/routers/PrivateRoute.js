import React from 'react'
import { Redirect, Route } from 'react-router-dom'
import PropTypes from 'prop-types'

export const PrivateRoute = ({ 
  isAuthenticated,
  component:Component, // se recibe component pero se renombra con C mayuscula para que se vea bien como componente
  ...rest // los demas argumentos como el exact y el path
}) => {
  return (
    <Route
    {...rest} // le paso el resto de los argumentos
    component={(props) => ( // aqui se renombra de rest a props
      (isAuthenticated) 
      ? (<Component {...props} />) 
      : (<Redirect to="/login" />)
    )} // llamo al componente con un callback para obtener sus props
    />
  )
}
PrivateRoute.propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
  component: PropTypes.func.isRequired,
}
