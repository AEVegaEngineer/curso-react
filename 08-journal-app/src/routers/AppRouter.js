import { getAuth } from '@firebase/auth'
import React, { useEffect, useState } from 'react'
import { useDispatch } from 'react-redux'
import { Route, BrowserRouter as Router, Switch, Redirect } from 'react-router-dom'
import { login } from '../actions/auth'
import { JournalScreen } from '../components/journal/JournalScreen'
import { AuthRouter } from './AuthRouter'

export const AppRouter = () => {

  // Mantener el estado de la autenticacion de firebase al recargar con un observable

  const dispatch = useDispatch();
  
  // checking  es un estado simple, la idea es mostrar un texto
  // de precarga de pantalla cuando se presiona F5, inicialmente
  // el valor de checking es siempre true, una vez se obtiene 
  // una respuesta del efecto, se setea checking en false y se
  // muestra la pagina.
  const [checking, setChecking] = useState(true);
  const [isLoggedIn, setIsLoggedIn] = useState(false);


  useEffect(() => {
    getAuth().onAuthStateChanged((user) => {
      if(user?.uid){
        dispatch(login(user.uid, user.displayName));
        setIsLoggedIn(true);
      } else { 
        setIsLoggedIn(false);
      }
      setChecking(false);
    });
  }, [dispatch,setChecking])

  
  if(checking) {
    return (
      <h1>Espere...</h1>
    )
  }

  return (    
    <Router> 
      <div>
        <Switch>
          <Route                       
            path="/auth" 
            component={AuthRouter}
          />   
          <Route
            exact             
            path="/" 
            component={JournalScreen}
          />
          {/* Cualquier otra ruta desconocida, redirige al /auth */}
          <Redirect to="/auth"/>
        </Switch>
      </div>     
      
    </Router>
  )
}

