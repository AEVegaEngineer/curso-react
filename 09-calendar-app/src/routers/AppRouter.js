import React from "react";
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Redirect,
} from "react-router-dom";
import { AuthScreen } from '../components/auth/AuthScreen';
import { CalendarScreen } from '../components/calendar/CalendarScreen';

export const AppRouter = () => {
  return (
    <Router> 
      <div>
        <Switch>
          <Route exact path="/auth" component={AuthScreen}/>
          <Route exact path="/" component={CalendarScreen}/>
          {/* Cualquier otra ruta desconocida, redirige al /auth */}
          <Redirect to="/auth"/>
        </Switch>
      </div>           
    </Router>
  )
}
