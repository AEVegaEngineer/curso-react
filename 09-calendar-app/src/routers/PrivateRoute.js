import React from 'react'
import { Redirect, Route } from 'react-router-dom'
import PropTypes from 'prop-types'

export const PrivateRoute = ({ 
  isAuthenticated,
  component:Component, // se recibe component pero se renombra con C mayuscula para que se vea bien como componente
  ...rest // los demas argumentos como el exact y el path
}) => {
  // obtenemos del rest, la ultima url visitada y la guardamos en localStorage para ingresar ahi la proxima vez que inicie sesion
  // adicionalmente, si la ruta es /search, le concatenamos la busqueda realizada

  //const loc = rest.location.pathname;
  //localStorage.setItem('lastPath', (loc === '/search') ? loc+rest.location.search : loc);

  return (
    <Route
    {...rest} // le paso el resto de los argumentos
    component={(props) => ( // aqui se renombra de rest a props
      (isAuthenticated) 
      ? (<Component {...props} />) 
      : (<Redirect to="/auth" />) // si no esta autenticado lo redirige al login
    )} // llamo al componente con un callback para obtener sus props
    />
  )
}
PrivateRoute.propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
  component: PropTypes.func.isRequired,
}
