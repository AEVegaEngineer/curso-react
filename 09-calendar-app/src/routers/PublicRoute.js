import React from 'react'
import { Redirect, Route } from 'react-router-dom'
import PropTypes from 'prop-types'

export const PublicRoute = ({ 
  isAuthenticated,
  component:Component, // se recibe component pero se renombra con C mayuscula para que se vea bien como componente
  ...rest // los demas argumentos como el exact y el path
}) => {
  return (
    <Route
    {...rest} // le paso el resto de los argumentos
    component={(props) => ( // aqui se renombra de rest a props
      (!isAuthenticated) 
      ? (<Component {...props} />) 
      : (<Redirect to="/" />) // si esta autenticado lo mantiene en el dashboard
    )} // llamo al componente con un callback para obtener sus props
    />
  )
}
PublicRoute.propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
  component: PropTypes.func.isRequired,
}
