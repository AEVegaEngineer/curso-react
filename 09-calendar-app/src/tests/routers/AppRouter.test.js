import React from 'react'
import { mount } from 'enzyme'
import { Provider } from 'react-redux'

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { AppRouter } from '../../routers/AppRouter';


const middlewares = [thunk];
const mockStore = configureStore(middlewares);



describe('Pruebas en <AppRouter/>', () => {
  test('Se debe mostrar correctamente', () => {
    const initState = {
      auth: {
        checking: true
      }
    };
    let store = mockStore(initState);
    
    const wrapper = mount(
      <Provider store={store}>
        <AppRouter/>
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('h5').exists()).toBe(true);
  });
});