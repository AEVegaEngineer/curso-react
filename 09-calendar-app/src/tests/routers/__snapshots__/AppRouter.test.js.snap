// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AppRouter/> Se debe mostrar correctamente 1`] = `
<h5>
  Espere...
</h5>
`;
